subxt/src/config/mod.rs

```diff
@@ -128,51 +128,24 @@ pub trait Header: Sized + Encode + Decode {
 #[cfg(feature = "substrate-compat")]
 mod substrate_impls {
     use super::*;
-    use primitive_types::{H256, U256};
 
-    impl<N, H> Header for sp_runtime::generic::Header<N, H>
+    impl<T: sp_runtime::traits::Header> Header for T
     where
-        Self: Encode + Decode,
-        N: Copy + Into<U256> + Into<u64> + TryFrom<U256>,
-        H: sp_runtime::traits::Hash + Hasher,
+        u64: From<<T as sp_runtime::traits::Header>::Number>,
```

Comment:
This essentially breaks things and only will work with block number types that implements Into<u64>?!

I don't understand the benefit with this except simplifying the trait a bit? 
Can you explain please?

However, I think the hashing stuff you added below simplifies things

Reply:
See subxt Header:
```rs
/// This represents the block header type used by a node.
pub trait Header: Sized + Encode + Decode {
    /// The block number type for this header.
    type Number: Into<u64>;
    /// The hasher used to hash this header.
    type Hasher: Hasher;

    /// Return the block number of this header.
    fn number(&self) -> Self::Number;

    /// Hash this header.
    fn hash(&self) -> <Self::Hasher as Hasher>::Output {
        Self::Hasher::hash_of(self)
    }
}
```

Reply:
I also do not like the `Into<u64>`, but subxt simplify related type defintions.

Reply:
Ok, I see it's probably better to change the number type on the header trait then.

/cc @lexnv @jsdw thoughts?

Reply:
I don't think that we should have a generic `impl<T: sp_runtime::traits::Header> Header for T` impl behind `substrate-compat`, because it would mean that enabling `substrate-compat` would potentially lead to breakages (and prevent any other impls from working).

Reply:
> I also do not like the Into<u64>, but subxt simplify related type defintions.

> Ok, I see it's probably better to change the number type on the header trait then.

Maybe I'm missing something; what would you propose changing about the number type? `Into<u64>` is just to allow us to return some valid numbers and compare them etc (it was a simplification over requiring a bunch of bounds, really; easier to just ask for a number, and prob it is always u32/u64 anyway?)